core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/Controller.scala

```diff
@@ -365,50 +365,74 @@ class Controller(
-        principalStates(p) = PrincipalState.Uninitialized
-        principalInCurrentStage.add(p)
-      }
-      workflow.startOperators.foreach { x =>
-        val principal = principalBiMap.get(x)
-        AdvancedMessageSending.nonBlockingAskWithRetry(
-          principal,
-          AckedPrincipalInitialization(Array()),
-          10,
-          0,
-          y =>
-            y match {
-              case AckWithInformation(z) =>
-                workflow.operators(x) = z.asInstanceOf[OpExecConfig]
-                //assign exception breakpoint before all breakpoints
-                if (!recoveryMode) {
-                  AdvancedMessageSending.blockingAskWithRetry(
-                    principal,
-                    AssignBreakpoint(
-                      new ExceptionGlobalBreakpoint(x.operator + "-ExceptionBreakpoint")
-                    ),
-                    10
+      try {
+        val nodes = availableNodes
+        log.info("start initialization --------cluster have " + nodes.length + " nodes---------")
+        for (k <- workflow.startOperators) {
+          val v = workflow.operators(k)
+          val p = context.actorOf(
+            Principal.props(v).withDeploy(Deploy(scope = RemoteScope(getPrincipalNode(nodes)))),
+            v.tag.operator
+          )
+          principalBiMap.put(k, p)
+          principalStates(p) = PrincipalState.Uninitialized
+          principalInCurrentStage.add(p)
+        }
+        workflow.startOperators.foreach { x =>
+          val principal = principalBiMap.get(x)
+          AdvancedMessageSending.nonBlockingAskWithRetry(
+            principal,
+            AckedPrincipalInitialization(Array()),
+            10,
+            0,
+            y =>
+              y match {
+                case AckWithInformation(z) =>
+                  workflow.operators(x) = z.asInstanceOf[OpExecConfig]
+                  //assign exception breakpoint before all breakpoints
+                  if (!recoveryMode) {
+                    AdvancedMessageSending.blockingAskWithRetry(
+                      principal,
+                      AssignBreakpoint(
+                        new ExceptionGlobalBreakpoint(x.operator + "-ExceptionBreakpoint")
+                      ),
+                      10
+                    )
+                  }
+                case other =>
+                  throw new WorkflowRuntimeException(
+                    WorkflowRuntimeError(
+                      "principal didn't return updated metadata",
+                      "Controller:receive:AckedControllerInitialization",
+                      Map()
+                    )
                   )
-                }
-              case other => throw new AmberException("principal didn't return updated metadata")
-            }
-        )
+              }
+          )
+        }
+        frontier ++= workflow.startOperators.flatMap(workflow.outLinks(_))
+      } catch {
+        case e: WorkflowRuntimeException =>
```

Comment:
I see an inconsistency in the error handling here. The controller is not using the `ErrorLogger` class but the pricipal is using it. It's good to uniformly have just one way to do it.